Prepare for release 131.1

basilgello opened the pull request:
@mborgerding @jtojnar Here I will push commits for review

basilgello commented:
I pushed the commit chain!

TODO to check / consider:

* https://github.com/mborgerding/kissfft/pull/58/files#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R188
* Whether or not we need pkg-config file `kissfft.pc`generated by Makefile and whether we can use `sed` there

Also, @mborgerding please check CMake `int32_t` and `int16_t` Python tests: both do 100% fail regardless of Python / NumPy version:

* https://github.com/mborgerding/kissfft/pull/58/files?file-filters%5B%5D=.c&file-filters%5B%5D=.h&file-filters%5B%5D=.sh&file-filters%5B%5D=.txt&file-filters%5B%5D=No+extension#diff-c4f91425d8a646ca04ec942f8d1854cf03f6ec09a5716d8904c25b8c740dec96R218
* https://github.com/mborgerding/kissfft/pull/58/files?file-filters%5B%5D=.c&file-filters%5B%5D=.h&file-filters%5B%5D=.sh&file-filters%5B%5D=.txt&file-filters%5B%5D=No+extension#diff-c4f91425d8a646ca04ec942f8d1854cf03f6ec09a5716d8904c25b8c740dec96R225
* https://github.com/mborgerding/kissfft/pull/58/files?file-filters%5B%5D=.c&file-filters%5B%5D=.h&file-filters%5B%5D=.sh&file-filters%5B%5D=.txt&file-filters%5B%5D=No+extension#diff-c4f91425d8a646ca04ec942f8d1854cf03f6ec09a5716d8904c25b8c740dec96R257
* https://github.com/mborgerding/kissfft/pull/58/files?file-filters%5B%5D=.c&file-filters%5B%5D=.h&file-filters%5B%5D=.sh&file-filters%5B%5D=.txt&file-filters%5B%5D=No+extension#diff-c4f91425d8a646ca04ec942f8d1854cf03f6ec09a5716d8904c25b8c740dec96R264

basilgello commented:
Later I will update documentation because i introduced many new Makefile and CMake configuration options.

jtojnar commented:
Yes the `.pc` file should be installed and respect the PREFIX and other variables passed to `make`. Using `sed` should be okay.

mborgerding commented:
Thanks for your work on this @basilgello @jtojnar .  I'll take a look at the changes and the failing int tests tonight.

mborgerding commented:
> Also, @mborgerding please check CMake `int32_t` and `int16_t` Python tests: both do 100% fail regardless of Python / NumPy version:

Can you provide commands that demonstrate this?  This works on my machine:
```
mkdir build
cd build
cmake ..
make
make test
```

basilgello commented:
int16_t:

mkdir build && cd build &&	 cmake -DKISSFFT_DATATYPE=int16_t .. && make && make test &&	cd .. && rm -rf build

int32_t:

mkdir build && cd build &&	 cmake -DKISSFFT_DATATYPE=int32_t .. && make && make test &&	cd .. && rm -rf build

basilgello commented:
Both int type tests fail inside test/testkiss.py : for int16_t, the fft_int16_t returns zero-dimensional array that is failed to transform and for int32_t, SNR calculated by fft_int32_t is less than the value calculated by numpy.

basilgello commented:
Interesting that running the same Python tests from Makefile succeeds :) I woke up with that fact bogging my mind, gonna check why CMake invocation fails.

basilgello commented:
> So here are still two problems: one is CMake not passing `DATATYPE` environment variable to `testkiss.py` and the issue with transforms itself. I will fix CMake now

Well, I was wrong on CMake problem. The `build/Testing/Temporary/LastTest.log` states:

```
8/8 Testing: testkiss.py
8/8 Test: testkiss.py
Command: "/usr/bin/python3.9" "/tmp/kissfft/test/testkiss.py"
Directory: /tmp/kissfft/build/test
"testkiss.py" start time: Jan 28 06:10 UTC
Output:
----------------------------------------------------------
Testing multi-dimensional FFTs
../tools/fft_int16_t -n 2
Traceback (most recent call last):
  File "/tmp/kissfft/test/testkiss.py", line 138, in <module>
    main()
  File "/tmp/kissfft/test/testkiss.py", line 134, in main
    test_fft(dim)
  File "/tmp/kissfft/test/testkiss.py", line 74, in test_fft
    x2 = dofft(x, do_real)
  File "/tmp/kissfft/test/testkiss.py", line 121, in dofft
    return np.reshape(res, dims)
  File "<__array_function__ internals>", line 5, in reshape
  File "/usr/lib/python3/dist-packages/numpy/core/fromnumeric.py", line 299, in reshape
    return _wrapfunc(a, 'reshape', newshape, order=order)
  File "/usr/lib/python3/dist-packages/numpy/core/fromnumeric.py", line 58, in _wrapfunc
    return bound(*args, **kwds)
ValueError: cannot reshape array of size 0 into shape (2,)
<end of output>
Test time =   0.22 sec
----------------------------------------------------------
Test Failed.
"testkiss.py" end time: Jan 28 06:10 UTC
"testkiss.py" time elapsed: 00:00:00
----------------------------------------------------------
```

I clearly see DATATYPE is passed here but nonetheless the error.

basilgello commented:
> Well, I was wrong on CMake problem. The `build/Testing/Temporary/LastTest.log` states:
> ...
> I clearly see DATATYPE is passed here but nonetheless the error.

It appears to be the issue with runtime path (RPATH). CMake strips RPATH from every executable (including `fft_int16_t`) on `make install`. This issue manifests for me only on `int16_t`.

basilgello commented:
I pushed the next revision that passes the extended testsuite.

@mborgerding @jtojnar I am going to push one more revision preparing pkg-config in Makefile and reflect that in testsuite & documentation.
If you want to run extended testsuite, invoke 'sh test/kissfft-testsuite.sh' from kissfft root directory and let it run for ~1h.

mborgerding commented:
The `kssfft-testsuite.sh` seems fine on an Ubuntu 20.04 system, but has inconsistent failures under CentOS 8:
```
[/tmp/kissfft-testsuite]$ grep 'error while' -r|sort -u
test-make-double-shared-tools:test/testcpp: error while loading shared libraries: libkissfft_double.so.131: cannot open shared object file: No such file or directory
test-make-float-shared-tools:test/testcpp: error while loading shared libraries: libkissfft_float.so.131: cannot open shared object file: No such file or directory
test-make-int16_t-shared-tools:test/testcpp: error while loading shared libraries: libkissfft_int16_t.so.131: cannot open shared object file: No such file or directory
test-make-int32_t-shared-tools:test/testcpp: error while loading shared libraries: libkissfft_int32_t.so.131: cannot open shared object file: No such file or directory
test-make-simd-shared-tools:test/testcpp: error while loading shared libraries: libkissfft_simd.so.131: cannot open shared object file: No such file or directory
```

basilgello commented:
Thanks for testing! I will spin a container with centos 6/7/8 and verify why it fails!

basilgello commented:
In the next update:

* `testsse` is invoked from `testall` only for `simd` data type
* `$LD_LIBRARY_PATH` escaped to `$(LD_LIBRARY_PATH)`
* `testcpp` is now a run-step like `testsse` to simplify testsuite
* testsuite now accepts custom output directory location

basilgello commented:
@mborgerding @jtojnar I finally pushed the release candidate.

In this release:

 * Make and CMake arguments are simplified and unified
 * Added Windows flags (tested with MSVC 2017)
 * Added extended testsuite
 * Added pkg-config support for Make
 * solib scheme changed from 'kissfft_${DATATYPE}' to 'kissfft-${DATATYPE}' and 'kissfft-${DATATYPE}-openmp', respectively.

basilgello commented:
If possible, please review before Feb 8th, so I can migrate 131.1.0 to Debian bullseye without too much hassle.

jtojnar commented on CMakeLists.txt:
```diff
@@ -1,149 +1,351 @@
+#
+
+if (CMAKE_INSTALL_INCLUDEDIR MATCHES "[\\/]?include[\\/]?$")
+    set(CMAKE_INSTALL_INCLUDEDIR "${CMAKE_INSTALL_INCLUDEDIR}/kissfft")
+    set(CMAKE_INSTALL_FULL_INCLUDEDIR "${CMAKE_INSTALL_FULL_INCLUDEDIR}/kissfft")
```
I find overriding the `GNUInstallDirs` suspect since it goes against expectations what those variables usually contain. It would be cleaner to use separate pair of variables (e.g. called `pkgincludedir` since `pkg` prefix is typically used for package-specific subdirectories of the directories with unprefixed variable names.

basilgello replied:
+1

jtojnar commented on kissfft.pc.in:
```diff
@@ -5,5 +5,5 @@ Name: kissfft
-Cflags: -I${includedir} @PKG_KISSFFT_DEFS@
\ No newline at end of file
+Libs: @PKG_OPENMP@ -L${libdir} -l@KISSFFT_OUTPUT_NAME@
+Cflags: -I${includedir} @PKG_KISSFFT_DEFS@
```
By convention `includedir` variable is the global one and `-I${includedir}/kissfft` would be used here. But I do not foresee any actual problems from not respecting the convention here.

basilgello replied:
Good point! Let me fix that in both Makefiles and CMakeLists

basilgello commented on Makefile:
```diff
@@ -1,58 +1,307 @@
+    _UNAME_ARCH = $(shell uname -i)
+
+    ifeq (,$(_UNAME_ARCH))
+	_UNAME_ARCH = $(shell uname -m)
```
I could not make nested `if` statement work with tabs... They have to be on the same tab-indentation level not to be interpreted as shell or rule commands. I cannot find the StackOverflow answer I referred to right now though...

basilgello commented on Makefile:
```diff
@@ -1,58 +1,307 @@
+#
+# -DKISS_FFT_BUILD is used for library artifacts, so
+# consumer executable in 'test' and 'tools' do _NOT_
+# need it. pkg-conffig output does not need it too
```
Yes, sure, thx for the spot-on!

jtojnar commented on CMakeLists.txt:
```diff
@@ -1,149 +1,351 @@
+# Declare CMAKE_INSTALL_{FULL_,}PKGINCLUDEDIR for kissfft include path
+#
+
+if (NOT DEFINED CMAKE_INSTALL_PKGINCLUDEDIR)
```
This can be made unconditional. And probably should avoid the `CMAKE_INSTALL_` prefix since it does not come from CMake.

basilgello replied:
I think `INSTALL_PKGINCLUDEDIR` is fair name reflecting that it does not come from CMake internals but also not coming from `PKGINCLUDEDIR` environment variable if specified. What do you think?

jtojnar replied:
Sounds okay to me too. Though I have not seen `PKGINCLUDEDIR` environment variable being used and I would not expect it to be a problem. Or do CMake or GNU Make just pick up random env vars?

basilgello replied:
> Or do CMake or GNU Make just pick up random env vars?

In CMake, one explicitly needs to ask `$ENV{NAME}`, GNU make picks all them and passes the `export`-ed ones to submakes.

jtojnar commented on kissfft.pc.in:
```diff
@@ -1,9 +1,10 @@
-Cflags: -I${includedir} @PKG_KISSFFT_DEFS@
\ No newline at end of file
+Libs: @PKG_OPENMP@ -L${libdir} -l@KISSFFT_OUTPUT_NAME@
+Cflags: -I${includedir} -I${pkgincludedir} @PKG_KISSFFT_DEFS@
```
This is the same as before but with a different variable name and even less idiomatic.

basilgello replied:
To be honest, I dont understand what is wrong here. If I specify `${includedir}/kissfft` here, I will have to suppress the ability to override `PKGINCLUDEDIR`. And I have to edit both `Makefile` and `CMakeLists.txt`...

jtojnar replied:
There should be no need to override `pkgincludedir`, people can already override the includedir.

The most idiomatic would be something like:

```patch
diff --git a/CMakeLists.txt b/CMakeLists.txt
index bb72a0f..f4cf632 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -106,14 +106,10 @@ if(CMAKE_SYSTEM_NAME MATCHES "^(Linux|kFreeBSD|GNU)$" AND NOT CMAKE_CROSSCOMPILI
 endif()
 
 #
-# Declare CMAKE_INSTALL_{FULL_,}PKGINCLUDEDIR for kissfft include path
+# Declare PKGINCLUDEDIR for kissfft include path
 #
 
-if (NOT DEFINED CMAKE_INSTALL_PKGINCLUDEDIR)
-    set(CMAKE_INSTALL_PKGINCLUDEDIR "${CMAKE_INSTALL_INCLUDEDIR}/kissfft")
-    set(CMAKE_INSTALL_FULL_PKGINCLUDEDIR "${CMAKE_INSTALL_FULL_INCLUDEDIR}/kissfft")
-    message(STATUS "CMAKE_INSTALL_PKGINCLUDEDIR is ${CMAKE_INSTALL_PKGINCLUDEDIR}")
-endif()
+set(PKGINCLUDEDIR "${CMAKE_INSTALL_INCLUDEDIR}/kissfft")
 
 #
 # Declare kissfft library ( libkissfft.a / libkissfft-${KISSFFT_DATATYPE}.so.${MAKEFILE_EXTRACTED_VERSION} )
@@ -128,7 +124,7 @@ add_library(kissfft
 
 target_include_directories(kissfft PUBLIC
     $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}>
-    $<INSTALL_INTERFACE:${CMAKE_INSTALL_PKGINCLUDEDIR}>)
+    $<INSTALL_INTERFACE:${PKGINCLUDEDIR}>)
 
 #
 # Set compile definitions based on datatype and additional support flags
@@ -286,7 +282,7 @@ install(FILES kiss_fft.h
               kiss_fftnd.h
               kiss_fftndr.h
               kiss_fftr.h
-        DESTINATION "${CMAKE_INSTALL_PKGINCLUDEDIR}")
+        DESTINATION "${PKGINCLUDEDIR}")
 
 set(KISSFFT_INSTALL_CMAKE "${CMAKE_INSTALL_LIBDIR}/cmake/${PROJECT_NAME}"
     CACHE FILEPATH "Install destination of kissfft cmake modules")
@@ -307,6 +303,11 @@ foreach(_def ${KISSFFT_COMPILE_DEFINITIONS})
     set(PKG_KISSFFT_DEFS "${PKG_KISSFFT_DEFS} -D${_def}")
 endforeach()
 if (KISSFFT_PKGCONFIG)
+    set(pkgincludedir_for_pkg_config "\${includedir}/kissfft")
+    set(prefix_for_pkg_config "${CMAKE_INSTALL_PREFIX}")
+    join_paths(libdir_for_pkg_config "\${prefix}" "${CMAKE_INSTALL_LIBDIR}")
+    join_paths(includedir_for_pkg_config "\${prefix}" "${CMAKE_INSTALL_INCLUDEDIR}")
+
     if(KISSFFT_DATATYPE MATCHES "^simd$")
        list(APPEND KISSFFT_COMPILE_DEFINITIONS USE_SIMD)
        if (NOT MSVC)
diff --git a/kissfft.pc.in b/kissfft.pc.in
index 3f130ad..322147a 100644
--- a/kissfft.pc.in
+++ b/kissfft.pc.in
@@ -1,10 +1,10 @@
-libdir=@CMAKE_INSTALL_FULL_LIBDIR@
-includedir=@CMAKE_INSTALL_FULL_INCLUDEDIR@
-pkgincludedir=@CMAKE_INSTALL_FULL_PKGINCLUDEDIR@
+prefix=@prefix_for_pkg_config@
+libdir=@libdir_for_pkg_config@
+includedir=@includedir_for_pkg_config@
 
 Name: kissfft
 Description: A Fast Fourier Transform (FFT) library that tries to Keep it Simple, Stupid
 Version: @kissfft_VERSION@
 
 Libs: @PKG_OPENMP@ -L${libdir} -l@KISSFFT_OUTPUT_NAME@
-Cflags: -I${includedir} -I${pkgincludedir} @PKG_KISSFFT_DEFS@
+Cflags: -I@pkgincludedir_for_pkg_config@ @PKG_KISSFFT_DEFS@
```

but you would need [`join_paths`](https://github.com/jtojnar/cmake-snips/blob/3cf4ad9c4901a5d116adcf2773cad7097a3580c1/CMakeScripts/JoinPaths.cmake) macro for older versions of CMake.

basilgello replied:
Thanks! Let me incorporate that diff, adjust Makefile counterpart & run the installation testsuite.

basilgello replied:
Done

jtojnar commented on Makefile:
```diff
@@ -1,58 +1,313 @@
+	PKGCONFIG_PKG_KISSFFT_DEFS = $(TYPEFLAGS)
+	PKGCONFIG_KISSFFT_PREFIX = $(ABS_PREFIX)
+	PKGCONFIG_KISSFFT_INCLUDEDIR = $${prefix}/include
+	PKGCONFIG_KISSFFT_LIBDIR = $${prefix}/lib
```
Hmm, this will be problematic when `CANDIDATE_LIBDIR` matches `lib64`. Which can be fixed by using `CANDIDATE_LIBDIR_NAME=lib64` and and `PKGCONFIG_KISSFFT_LIBDIR = $${prefix}/$(CANDIDATE_LIBDIR_NAME)`.

But it gets even worse when `LIBDIR` is overridden (e.g. debian uses [`--libdir=\$${prefix}/lib/$(ARCH)`](https://wiki.debian.org/Multiarch/Implementation)). In the absence of `libdir` variable relative to `prefix` like CMake’s `GNUInstallDirs` module provides us, maybe something like

```make
ifeq "$(LIBDIR)" "$(PREFIX)/$(CANDIDATE_LIBDIR_NAME)"
 	PKGCONFIG_KISSFFT_PKGINCLUDEDIR = $${prefix}/$(CANDIDATE_LIBDIR_NAME)
else
 	PKGCONFIG_KISSFFT_PKGINCLUDEDIR = $(LIBDIR)
endif
```

would be good enough substitute.

basilgello replied:
Should I also substritute `$(INCDIR)` too for the same reason? I guess so. $(`BINDIR)` does not land in pkg-config file so it should be left intaxt.

jtojnar replied:
Yes, the `includedir` suffers from the same issue (sans candidate selection). We actually change that one on Nix to a different prefix.

basilgello replied:
Uploaded the fix + fixed `$(FFTWLIB)` in `test/Makefile`. Hopefully this time it is OK :)

jtojnar commented on Makefile:
```diff
@@ -1,58 +1,323 @@
+	PKGCONFIG_KISSFFT_OUTPUT_NAME = $(KISSFFTLIB_SHORTNAME)
+	PKGCONFIG_PKG_KISSFFT_DEFS = $(TYPEFLAGS)
+	PKGCONFIG_KISSFFT_PREFIX = $(ABS_PREFIX)
+  ifeq "$(ABS_INCLUDEDIR)" $(abspath "$(PREFIX)/include")
```
Hmm, getting the following error now:

Makefile:161: *** invalid syntax in conditional.  Stop.

jtojnar commented on Makefile:
```diff
@@ -1,58 +1,323 @@
+	PKGCONFIG_KISSFFT_OUTPUT_NAME = $(KISSFFTLIB_SHORTNAME)
+	PKGCONFIG_PKG_KISSFFT_DEFS = $(TYPEFLAGS)
+	PKGCONFIG_KISSFFT_PREFIX = $(ABS_PREFIX)
+  ifeq "$(ABS_INCLUDEDIR)" $(abspath "$(PREFIX)/include")
```
The following seems to work.

```patch
diff --git a/Makefile b/Makefile
index 2db2b90..7b4f2c2 100644
--- a/Makefile
+++ b/Makefile
@@ -158,13 +158,13 @@ ifneq ($(shell uname -s),Darwin)
 	PKGCONFIG_KISSFFT_OUTPUT_NAME = $(KISSFFTLIB_SHORTNAME)
 	PKGCONFIG_PKG_KISSFFT_DEFS = $(TYPEFLAGS)
 	PKGCONFIG_KISSFFT_PREFIX = $(ABS_PREFIX)
-  ifeq "$(ABS_INCLUDEDIR)" $(abspath "$(PREFIX)/include")
+  ifeq ($(ABS_INCLUDEDIR),$(PREFIX)/include)
 	PKGCONFIG_KISSFFT_INCLUDEDIR = $${prefix}/include
   else
 	PKGCONFIG_KISSFFT_INCLUDEDIR = $(ABS_INCLUDEDIR)
 
   endif
-  ifeq "$(ABS_LIBDIR)" $(abspath "$(PREFIX)/$(CANDIDATE_LIBDIR_NAME)")
+  ifeq ($(ABS_LIBDIR),$(PREFIX)/$(CANDIDATE_LIBDIR_NAME))
 	PKGCONFIG_KISSFFT_LIBDIR = $${prefix}/$(CANDIDATE_LIBDIR_NAME)
   else
 	PKGCONFIG_KISSFFT_LIBDIR = $(ABS_LIBDIR)
```

Had to remove abspath because it seems to always prepend `$PWD`, even when the path is already absolute, resulting in paths like `/build/kissfft//nix/store/7l8pzpb9irz2q07wscrysp27z53ya0xv-kissfft-131/include`.

basilgello replied:
Then, probably:

```Make
ifeq ($(ABS_INCLUDEDIR),$(ABS_PREFIX)/include
```

jtojnar replied:
Actually, the issue was not `abspath` itself but the quotes within. If I remove them, it works okay.

```patch
diff --git a/Makefile b/Makefile
index 2db2b90..3a5a5ac 100644
--- a/Makefile
+++ b/Makefile
@@ -158,13 +158,12 @@ ifneq ($(shell uname -s),Darwin)
 	PKGCONFIG_KISSFFT_OUTPUT_NAME = $(KISSFFTLIB_SHORTNAME)
 	PKGCONFIG_PKG_KISSFFT_DEFS = $(TYPEFLAGS)
 	PKGCONFIG_KISSFFT_PREFIX = $(ABS_PREFIX)
-  ifeq "$(ABS_INCLUDEDIR)" $(abspath "$(PREFIX)/include")
+  ifeq ($(ABS_INCLUDEDIR),$(abspath $(PREFIX)/include))
 	PKGCONFIG_KISSFFT_INCLUDEDIR = $${prefix}/include
   else
 	PKGCONFIG_KISSFFT_INCLUDEDIR = $(ABS_INCLUDEDIR)
-
   endif
-  ifeq "$(ABS_LIBDIR)" $(abspath "$(PREFIX)/$(CANDIDATE_LIBDIR_NAME)")
+  ifeq ($(ABS_LIBDIR),$(abspath $(PREFIX)/$(CANDIDATE_LIBDIR_NAME)))
 	PKGCONFIG_KISSFFT_LIBDIR = $${prefix}/$(CANDIDATE_LIBDIR_NAME)
   else
 	PKGCONFIG_KISSFFT_LIBDIR = $(ABS_LIBDIR)
```

basilgello replied:
> Then, probably:

This works for me on Debian sid & CentOS 8 chroots.

basilgello replied:
Pushed the fix :) `ABS_*` variables are already absolute paths so it should be equal to previous solution with `abspath` in place

jtojnar commented on Makefile:
```diff
@@ -1,58 +1,323 @@
+BINDIR ?= $(ABS_PREFIX)/bin
+export ABS_BINDIR = $(abspath $(BINDIR))
+
+INCDIR ?= $(ABS_PREFIX)/include
```
Maybe this could be renamed to `includedir` to better match the GNU variable names.

basilgello replied:
Then also `bindir`, `includedir`, `libdir` and `pkgincludedir`. And `ABS_` or `abs_` ? :)

jtojnar replied:
I think the casing is fine for makefiles (with pure make, the capital casing is slightly more common, in my experience). I mostly meant inc → include.

basilgello replied:
OK, pushed the fix replacing `INCDIR` witn `INCLUDEDIR` :)

jtojnar reviewed (approved):
The pkg-config files seem to work great with both CMake and GNU Make on Nix, both in a single prefix and different prefixes for devel files. Great job.

basilgello commented:
> The pkg-config files seem to work great with both CMake and GNU Make on Nix, both in a single prefix and different prefixes for devel files. Great job.

Thanks for improvements! :)

mborgerding commented:
Thanks for all the work @basilgello @jtojnar!  I'll look at merging this to master tonight or tomorrow.

mborgerding commented:
* I'm leaning towards removing the `kissfft_i32.h` and maybe `kfc.h` from the installation. These are less generally useful (and  less well tested).   Developers who want that stuff can build it themselves.
* I plan to remove README (having been replaced by README.md)

What do you think about an aggregation header `"${CMAKE_INSTALL_INCLUDEDIR}/kissfft.h"?
Something like
```c
/*Force compatibility with the installed shared lib*/
#define kiss_fft_scalar float

/*basic C interface for mixed radix complex FFTs*/
#include <kissfft/kiss_fft.h>

/*real 1D FFTs */
#include <kissfft/kiss_fftr.h>

/*multidimensional complex*/
#include <kissfft/kiss_fftnd.h>

/*multidimensional real*/
#include <kissfft/kiss_fftndr.h>

#ifdef __cplusplus
// header-only C++ version 
#  include <kissfft/kissfft.hh>
#endif
```

Advantages:

* provides a uniform and easy way for developers to gain access to all the different functions (the current naming is a mess!).
* enforces that data type version

mborgerding commented:
I'd recommend  8f47a67f595a6641c566087bf5277034be64f24d  which removes kfc.h, kissfft_i32.hh, and README as discussed above.

basilgello commented:
Thanks for merging! Sorry for delay on reviewing the kfc.h and kissfft_i32.hh removal!

I do like idea of an aggregate kissfft.h. Removing kfc.h from installation is fine without ABI bump because the solib still provides the necessary functions.

Further step is (maybe) to rename public API functions to be able to use several datatypes within a single application, but it is a breaking change that will need some discussion & preparation.

basilgello commented:
Also please create a tagged release whenever KFVER constants change. This is needed for distribution watchdogs like Debian's vcswatch to track the availability of new version automatically.

For example, see https://github.com/kodi-pvr/pvr.iptvsimple
